### PR TITLE
[Regression] Fix usage of FastlanePty.spawn()

### PR DESCRIPTION
**Pull request by robinkunde**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

99e486a6987b84bcd64db8670cf68c2c9fb644ab introduced FastlanePty.spawn() to provide a `PTY` for Windows. In caused several regressions that don't seem to be covered by tests.
1) It replaced PTY.spawn() which takes a string as its first parameter. FastlanePty.spawn() converts all parameters into an array for Open3.popen2e() but doesn't convert it back to string for PTY.spawn().
2) The old PTY.spawn() call sites had swapped stdout and stdin, even going to far as read from "stdin" (actually stdout). This was then carried forward, but because PTY.spawn() and Open3.popen2e() provide these sockets in a different order, calls to FastlanePty.spawn() don't work as intended on Mac OS.

### Description

This PR fixed these issues and also clears up the incorrect labelling at all call sites.

### CODE REVIEW

**Comment by robinkunde** [comment]
I signed the CLA.

**Comment by janpio** [comment]
Wow, glad you caught that. 

Explains why these params never really made sense to me there, but I just went with it as the tests seemed to be happy (and I don't really understand what is happening there anyway. Even googling now it seems the examples that uss the naming `stdin`/`stdout` are split 50/50 on the param order...).

Do you understand all that stuff well enough so you could maybe add some tests covering this?

**Comment by robinkunde** [comment]
I don’t have a good enough grip on the test system to attempt that yet, and
won’t have room in my schedule to learn it any time soon. Sorry!

File: fastlane_core/lib/fastlane_core/itunes_transporter.rb
Comment (lacostej) [inline fastlane_core/lib/fastlane_core/itunes_transporter.rb]: When dealing with stdin/stdout and pipes, the naming is always confusing. The command's stdout is our stdin, and vice versa.

I think it's why the ruby API uses `r` and `w` for the `spawn` call (which we are wrapping for cross-platform reasons).

So we have to be clear at what we are naming here. Originally the first parameter was named `stdin` as it provided input to the block. Now it changes to `stdout` which means the command's stdout.

From a programming perspective what do we expect? Do we want to keep matching the spawn call's parameters? Do we want to name the parameter from the command or the block perspective?

**Review by lacostej — COMMENTED** [review]
Commenting on the historical reason for the parameters namings.

**Comment by lacostej** [comment]
Hello @robinkunde  thanks for catching the regression.

Actually the main issues are here https://github.com/fastlane/fastlane/blob/99e486a6987b84bcd64db8670cf68c2c9fb644ab/fastlane_core/lib/fastlane_core/fastlane_pty.rb#L8-L9

1. the parameters got inversed. It wasn't meant to.
2. the splat operator was introduced at the method level.

So it could be fixed by just swapping the parameters on the second line and removing the splat operator. I don't see why it was introduced and why FastlanePty has now a very different API than the original spawn. 

I completely missed those changes in the original review. :(

An implementation closer to the original spawn could be:
```
  class FastlanePty
    def self.spawn(command, &block)
      require 'pty'
      PTY.spawn(command) do |stdout, stdin, pid|
        block.call(stdout, stdin, pid)
      end
    rescue LoadError
      require 'open3'
      Open3.popen2e(*command) do |r, w, p|
        yield(w, r, p.value.pid) # note the inversion

        r.close
        w.close
        p.value.exitstatus
      end
    end
  end
```

I submitted #11691 to test the implementation on Windows. But I don't intend to merge it.

As for the inverted naming I explained it above. I am a bit uncertain of what we want to do here. It's not necessary for the fix though. I am +0 on that particular change.

**Comment by janpio** [comment]
> 1. the parameters got inversed. It wasn't meant to.
> 2. the splat operator was introduced at the method level.
>
> I don't see why it was introduced and why FastlanePty has now a very different API than the original spawn.

100% bad copy paste by me as this is a combination of an old PR, external code and review comments that I combined into one (ugly) package.

How about we just standardize on the naming from the ruby docs, `r, w, pid`?
https://ruby-doc.org/stdlib-2.4.0/libdoc/pty/rdoc/PTY.html#method-c-getpty-label-Return+values

**Comment by lacostej** [comment]
> How about we just standardize on the naming from the ruby docs, r, w, pid?
> https://ruby-doc.org/stdlib-2.4.0/libdoc/pty/rdoc/PTY.html#method-c-getpty-label-Return+values

Either that or make them more explicit. E.g. `commandStdout` and `commandStdin` or similar

**Comment by robinkunde** [comment]
+1 on making it more explicit. I can also remove the splat and amend my PR.

**Comment by robinkunde** [comment]
PR is updated. I removed the splat from FastlanePty.spawn()'s method signature but did not add it to the call to Open3.popen2e(). According to [the docs](https://ruby-doc.org/stdlib-1.9.3/libdoc/open3/rdoc/Open3.html#method-c-popen2e), it's already in the method signature, so it shouldn't be necessary at the call site. Please let me know if I'm wrong.

**Comment by janpio** [comment]
Rubocop unfortunately doesn't like the variable names:

```
[19:47:17]: ▸ Offenses:
[19:47:17]: ▸ snapshot/lib/snapshot/screenshot_rotate.rb:31:54: C: Style/VariableName: Use snake_case for variable names.
[19:47:17]: ▸ FastlaneCore::FastlanePty.spawn(command) do |commandStdout, commandStdin, pid|
[19:47:17]: ▸ ^^^^^^^^^^^^^
[19:47:17]: ▸ snapshot/lib/snapshot/screenshot_rotate.rb:31:69: C: Style/VariableName: Use snake_case for variable names.
[19:47:17]: ▸ FastlaneCore::FastlanePty.spawn(command) do |commandStdout, commandStdin, pid|
[19:47:17]: ▸ ^^^^^^^^^^^^
[19:47:17]: ▸ fastlane_core/lib/fastlane_core/command_executor.rb:50:56: C: Style/VariableName: Use snake_case for variable names.
[19:47:17]: ▸ FastlaneCore::FastlanePty.spawn(command) do |commandStdout, commandStdin, pid|
[19:47:17]: ▸ ^^^^^^^^^^^^^
[19:47:17]: ▸ fastlane_core/lib/fastlane_core/command_executor.rb:50:71: C: Style/VariableName: Use snake_case for variable names.
[19:47:17]: ▸ FastlaneCore::FastlanePty.spawn(command) do |commandStdout, commandStdin, pid|
[19:47:17]: ▸ ^^^^^^^^^^^^
[19:47:17]: ▸ fastlane_core/lib/fastlane_core/itunes_transporter.rb:47:54: C: Style/VariableName: Use snake_case for variable names.
[19:47:17]: ▸ FastlaneCore::FastlanePty.spawn(command) do |commandStdout, commandStdin, pid|
[19:47:17]: ▸ ^^^^^^^^^^^^^
[19:47:17]: ▸ fastlane_core/lib/fastlane_core/itunes_transporter.rb:47:69: C: Style/VariableName: Use snake_case for variable names.
[19:47:17]: ▸ FastlaneCore::FastlanePty.spawn(command) do |commandStdout, commandStdin, pid|
[19:47:17]: ▸ ^^^^^^^^^^^^
[19:47:17]: ▸ fastlane_core/lib/fastlane_core/fastlane_pty.rb:8:30: C: Style/VariableName: Use snake_case for variable names.
[19:47:17]: ▸ PTY.spawn(command) do |commandStdout, commandStdin, pid|
[19:47:17]: ▸ ^^^^^^^^^^^^^
[19:47:17]: ▸ fastlane_core/lib/fastlane_core/fastlane_pty.rb:8:45: C: Style/VariableName: Use snake_case for variable names.
[19:47:17]: ▸ PTY.spawn(command) do |commandStdout, commandStdin, pid|
[19:47:17]: ▸ ^^^^^^^^^^^^
[19:47:17]: ▸ fastlane_core/lib/fastlane_core/fastlane_pty.rb:13:34: C: Style/VariableName: Use snake_case for variable names.
[19:47:17]: ▸ Open3.popen2e(command) do |commandStdin, commandStdout, p| # note the inversion
[19:47:17]: ▸ ^^^^^^^^^^^^
[19:47:17]: ▸ fastlane_core/lib/fastlane_core/fastlane_pty.rb:13:48: C: Style/VariableName: Use snake_case for variable names.
[19:47:17]: ▸ Open3.popen2e(command) do |commandStdin, commandStdout, p| # note the inversion
[19:47:17]: ▸ ^^^^^^^^^^^^^
```

**Comment by lacostej** [comment]
Ooops my bad. Switching between languages all the time and I make those kind of mistakes.

Here's a script that will quickly fix it on your branch:

```
git grep commandStd | cut -d ':' -f 1 | xargs -L1 sed -e 's/commandStd/command_std/g' -i '' 
git commit -am "Rename variables to please rubocop"
git push
```

**Review by lacostej — APPROVED** [review]

**Comment by lacostej** [comment]
Approved once the rubocop fixes are merged of course.

**Comment by robinkunde** [comment]
Done!

**Comment by janpio** [comment]
Thanks so much @robinkunde for fixing my mess - I really hope to not see PTY now for quite some time ;) 
Thanks as always to @lacostej who actually knows some ruby :)

💯

**Pull request merged**
